src/output.c

```diff
@@ -0,0 +1,308 @@
+            case OUT_H1:
+                set_led(LED_STATH1); break;
+            case OUT_L0:
+                set_led(LED_STATL0); break;
+            case OUT_L1:
+                set_led(LED_STATL1); break;
+            case OUT_L2:
+                set_led(LED_STATL2); break;
+            case OUT_L3:
+                set_led(LED_STATL3); break;
+            default: break;
+        }
+    } else {
+        output_inhibited[out] = false;
+        // clear channel error LED
+        switch (out) {
+            case OUT_H0:
+                clear_led(LED_STATH0); break;
+            case OUT_H1:
+                clear_led(LED_STATH1); break;
+            case OUT_L0:
+                clear_led(LED_STATL0); break;
+            case OUT_L1:
+                clear_led(LED_STATL1); break;
+            case OUT_L2:
+                clear_led(LED_STATL2); break;
+            case OUT_L3:
+                clear_led(LED_STATL3); break;
+            default: break;
+        }
+    }
+}
+
+void handle_uvlo(void) {
+    // Test if global voltage is below 10.2V
+    if ((battery.success) && (battery.voltage < 10200)) {
+        uvlo_delay++;
+        if (uvlo_delay > UVLO_DELAY) {
+            disable_all_outputs(true);
+            set_led(LED_FLAT);
+
+            // Disable systick & USB
+            systick_counter_disable();
+            usb_deinit();
+
+            // Disable fan
+            fan_enable(false);
+
+            while (1) {
+                // flash flat LED
+                toggle_led(LED_FLAT);
+
+                for (unsigned int j = 0; j < 25; j++) {
+                    delay(20);
+                    iwdg_reset();
+                }
+            }
+        }
+    } else {
+        uvlo_delay = 0;
+    }
+}
+
+void detect_overcurrent(void) {
+    // Test individual output currents
+    for (output_t out=OUT_H0; out <= OUT_L3; out++) {
+        uint16_t current_limit = (out <= OUT_H1) ? 20000 : 10000;
+
+        if (output_current[out] > current_limit) {
+            overcurrent_delay[out]++;
+            if (overcurrent_delay[out] > ADC_OVERCURRENT_DELAY) {
+                // disable channel
+                set_overcurrent(out, true);
+            }
+        } else {
+            overcurrent_delay[out] = 0;
+        }
+    }
+    if ((reg_5v.success) && (reg_5v.current > 2000)) {
+        overcurrent_delay[OUT_5V]++;
+        if (overcurrent_delay[OUT_5V] > REG_OVERCURRENT_DELAY) {
+            // disable channel
+            set_overcurrent(OUT_5V, true);
+        }
+    } else {
+        overcurrent_delay[OUT_5V] = 0;
+    }
+
+    // Test global current
+    // Also test combined output current
+    int total_current = 0;
+    for (output_t out=OUT_H0; out < OUT_5V; out++) {
+        total_current += output_current[out];
+    }
+    if (reg_5v.success) {
+        total_current += reg_5v.current;
+    }
+    if (
+        (total_current > 30000)
+        || ((battery.success) && (battery.current > 30000))
```

Comment:
You should also consider negative battery over current too (with a suitable threshold and delay). One of the bugs with the old firmware was that a negative battery current would break stuff (since the code assumed that the battery current would always be positive). By the looks of things that's been fixed, but it would be good to make use of the fact that negative currents can now be measured to enforce a negative limit.

Negative current occurs with particularly massive robots when they stop or change direction rapidly. The kinetic energy is dumped back into the battery and the power can exceed the total consumption of the rest of the kit. Personally I reckon that 4A limit for a couple of seconds is fine.

Reply:
I've added a reverse current cut-off. It's currently set to -1A for 1 second because I'm concerned that uncontrolled reverse current for any period is very bad for the battery. There is an ability to adjust the delay in the field if that turns out to be an inappropriate value.